File: backend/src/pomelo_orbit/interfaces/api/dto/ci.py
"""CI 模块 DTO"""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel

# ── Credential ────────────────────────────────────────────────────────────────


class CredentialResp(BaseModel):
    id: str
    name: str
    type: str
    created_at: datetime

    model_config = {"from_attributes": True}


class CredentialCreateReq(BaseModel):
    name: str
    type: Literal["git_ssh", "git_token", "registry_token"]
    data: str


class CredentialUpdateReq(BaseModel):
    name: str | None = None
    data: str | None = None


# ── Stage ─────────────────────────────────────────────────────────────────────


class CheckoutConfigDto(BaseModel):
    ref: str = "{{ DEFAULT_BRANCH }}"


class DockerBuildConfigDto(BaseModel):
    context: str = "."
    dockerfile: str = "Dockerfile"
    image_name: str


class UnitTestConfigDto(BaseModel):
    image: str
    commands: list[str]
    artifact_paths: list[str] = []


class StageDefinitionDto(BaseModel):
    name: str
    type: Literal["checkout", "docker_build", "unit_test", "custom"]
    depends_on: list[str] = []
    config: CheckoutConfigDto | DockerBuildConfigDto | UnitTestConfigDto | None = None
    steps: list[dict[str, Any]] | None = None  # custom 类型的 StepDefinition


# ── VariableDeclaration ───────────────────────────────────────────────────────


class VariableDeclarationDto(BaseModel):
    name: str
    description: str = ""
    required: bool = False
    default: Any = None
    secret: bool = False
    locked: bool = False


# ── PipelineTemplate ──────────────────────────────────────────────────────────


class PipelineTemplateResp(BaseModel):
    id: str
    name: str
    description: str
    stages: list[StageDefinitionDto]
    variable_declarations: list[VariableDeclarationDto]
    is_builtin: bool
    latest_snapshot_version: int | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PipelineTemplateCreateReq(BaseModel):
    name: str
    description: str = ""
    stages: list[StageDefinitionDto] = []
    variable_declarations: list[VariableDeclarationDto] = []


class PipelineTemplateUpdateReq(BaseModel):
    name: str | None = None
    description: str | None = None
    stages: list[StageDefinitionDto] | None = None
    variable_declarations: list[VariableDeclarationDto] | None = None


# ── PipelineSnapshot ──────────────────────────────────────────────────────────


class PipelineSnapshotListItemResp(BaseModel):
    id: str
    template_id: str
    version: int
    created_at: datetime

    model_config = {"from_attributes": True}


class PipelineSnapshotResp(BaseModel):
    id: str
    template_id: str
    version: int
    stages_snapshot: list[StageDefinitionDto]
    variable_declarations_snapshot: list[VariableDeclarationDto]
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Project ───────────────────────────────────────────────────────────────────


class ProjectResp(BaseModel):
    id: str
    name: str
    repository_url: str
    pipeline_snapshot_id: str
    git_credential_id: str | None
    variable_overrides: dict[str, Any]
    default_branch: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProjectCreateReq(BaseModel):
    name: str
    repository_url: str
    pipeline_snapshot_id: str
    git_credential_id: str | None = None
    variable_overrides: dict[str, Any] = {}
    default_branch: str = "master"


class ProjectUpdateReq(BaseModel):
    name: str | None = None
    repository_url: str | None = None
    pipeline_snapshot_id: str | None = None
    git_credential_id: str | None = None
    variable_overrides: dict[str, Any] | None = None
    default_branch: str | None = None


# ── PipelineRun ───────────────────────────────────────────────────────────────


class PipelineRunResp(BaseModel):
    id: str
    project_id: str
    pipeline_snapshot_id: str
    trigger: str
    trigger_ref: str
    status: str
    retry_of: str | None = None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TriggerPipelineReq(BaseModel):
    trigger_ref: str = ""
    variables: dict[str, Any] = {}


# ── Job ───────────────────────────────────────────────────────────────────────


class JobResp(BaseModel):
    id: str
    pipeline_run_id: str
    name: str
    status: str
    parent_job_id: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error_message: str | None = None

    model_config = {"from_attributes": True}


class JobLogResp(BaseModel):
    id: str
    job_id: str
    content: str
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Artifact ──────────────────────────────────────────────────────────────────


class ArtifactResp(BaseModel):
    id: str
    pipeline_run_id: str
    job_name: str
    type: str
    name: str
    path: str | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: internal/model/ci.go
package model

import "time"

type Credential struct {
	Id            string    `db:"id"`
	ProjectId     *string   `db:"project_id"`
	Name          string    `db:"name"`
	Type          string    `db:"type"`
	EncryptedData string    `db:"encrypted_data"`
	CreatedAt     time.Time `db:"created_at"`
}

type PipelineTemplate struct {
	Id                   string    `db:"id"`
	ProjectId            *string   `db:"project_id"`
	Name                 string    `db:"name"`
	Description          string    `db:"description"`
	VariableDeclarations string    `db:"variable_declarations"`
	Version              int       `db:"version"`
	CreatedAt            time.Time `db:"created_at"`
	UpdatedAt            time.Time `db:"updated_at"`
}

type BuildStage struct {
	Id          string    `db:"id"`
	ProjectId   *string   `db:"project_id"`
	Name        string    `db:"name"`
	Image       string    `db:"image"`
	Script      string    `db:"script"`
	Artifacts   *string   `db:"artifacts"`
	Description string    `db:"description"`
	Version     int       `db:"version"`
	CreatedAt   time.Time `db:"created_at"`
	UpdatedAt   time.Time `db:"updated_at"`
}

type PipelineTemplateStage struct {
	Id           string `db:"id"`
	TemplateId   string `db:"template_id"`
	StageId      string `db:"stage_id"`
	StageName    string `db:"stage_name"`
	StageVersion int    `db:"stage_version"`
	DependsOn    string `db:"depends_on"`
	SortOrder    int    `db:"sort_order"`
}

type PipelineSnapshot struct {
	Id                string    `db:"id"`
	ProjectId         *string   `db:"project_id"`
	TemplateId        string    `db:"template_id"`
	Version           int       `db:"version"`
	StagesSnapshot    string    `db:"stages_snapshot"`
	VariablesSnapshot string    `db:"variables_snapshot"`
	CreatedAt         time.Time `db:"created_at"`
}

type Repository struct {
	Id                string    `db:"id"`
	ProjectId         *string   `db:"project_id"`
	Name              string    `db:"name"`
	Code              string    `db:"code"`
	RepositoryURL     string    `db:"repository_url"`
	GitCredentialId   *string   `db:"git_credential_id"`
	VariableOverrides string    `db:"variable_overrides"`
	DefaultBranch     string    `db:"default_branch"`
	CreatedAt         time.Time `db:"created_at"`
	UpdatedAt         time.Time `db:"updated_at"`
}

type RepositoryWebhook struct {
	Id              string    `db:"id"`
	RepositoryId    string    `db:"repository_id"`
	Name            string    `db:"name"`
	TemplateId      string    `db:"template_id"`
	BranchFilter    *string   `db:"branch_filter"`
	EncryptedSecret string    `db:"encrypted_secret"`
	Enabled         bool      `db:"enabled"`
	CreatedAt       time.Time `db:"created_at"`
	UpdatedAt       time.Time `db:"updated_at"`
}

type PipelineRun struct {
	Id                string     `db:"id"`
	ProjectId         *string    `db:"project_id"`
	RepositoryId      string     `db:"repository_id"`
	RepositoryName    string     `db:"repository_name"`
	SnapshotId        string     `db:"snapshot_id"`
	TemplateId        string     `db:"template_id"`
	TemplateName      string     `db:"template_name"`
	TemplateVersion   int        `db:"template_version"`
	Trigger           string     `db:"trigger"`
	TriggerRef        string     `db:"trigger_ref"`
	VariablesSnapshot string     `db:"variables_snapshot"`
	Status            string     `db:"status"`
	RetryOf           *string    `db:"retry_of"`
	StartedAt         *time.Time `db:"started_at"`
	FinishedAt        *time.Time `db:"finished_at"`
	ErrorMessage      *string    `db:"error_message"`
	CreatedAt         time.Time  `db:"created_at"`
}

type StageRun struct {
	Id            string     `db:"id"`
	PipelineRunId string     `db:"pipeline_run_id"`
	StageId       string     `db:"stage_id"`
	StageName     string     `db:"stage_name"`
	Status        string     `db:"status"`
	StartedAt     *time.Time `db:"started_at"`
	FinishedAt    *time.Time `db:"finished_at"`
	ExitCode      *int       `db:"exit_code"`
	ErrorMessage  *string    `db:"error_message"`
}

type Artifact struct {
	Id             string    `db:"id"`
	ProjectId      *string   `db:"project_id"`
	PipelineRunId  string    `db:"pipeline_run_id"`
	RepositoryId   string    `db:"repository_id"`
	RepositoryName string    `db:"repository_name"`
	TemplateId     string    `db:"template_id"`
	TemplateName   string    `db:"template_name"`
	StageName      string    `db:"stage_name"`
	Type           string    `db:"type"`
	Name           string    `db:"name"`
	Path           *string   `db:"path"`
	CreatedAt      time.Time `db:"created_at"`
}

type StageDefinition struct {
	Name      string           `json:"name"`
	Id        string           `json:"id"`
	Image     string           `json:"image"`
	Version   int              `json:"version"`
	DependsOn []string         `json:"depends_on"`
	Script    string           `json:"script"`
	Artifacts []ArtifactConfig `json:"artifacts"`
}

type ArtifactConfig struct {
	Type string `json:"type"`
	Path string `json:"path"`
	Name string `json:"name"`
}

type VariableDeclaration struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Default     any    `json:"default"`
	Value       any    `json:"value"`
	Secret      bool   `json:"secret"`
	Source      string `json:"source"`
	Editable    bool   `json:"editable"`
}
